rainier-core/src/main/scala/com/stripe/rainier/sampler/LeapFrog.scala

```diff
@@ -60,6 +61,103 @@ private[sampler] case class LeapFrog(density: DensityFunction) {
     pqBuf(potentialIndex) = density.density * -1
   }
 
+  /**
+    * Perform l leapfrog steps updating pqBuf
+    * @param l the total number of leapfrog steps to perform
+    * @param stepSize the current step size
+    */
+  private def steps(l: Int, stepSize: Double): Unit = {
+    initialHalfThenFullStep(stepSize)
+    var i = 1
+    while (i < l) {
+      twoFullSteps(stepSize)
+      i += 1
+    }
+    finalHalfStep(stepSize)
+  }
+
+  /**
+    * Determine if a u-turn has been performed by checking the initial
+    * parameters against pqBuf
+    */
+  private def isUTurn(params: Array[Double]): Boolean = {
+
+    var out = 0.0
+    var i = 0
+    while (i < nVars) {
+      out += (pqBuf(i + nVars) - params(i + nVars)) * pqBuf(i)
+      i += 1
+    }
+
+    if (out.isNaN)
+      true
+    else
+      out < 0
+  }
+
+  /**
+    * Advance l0 steps and return the value
+    * of the longest-step size until a u-turn
+    */
+  private def longestStepUntilUTurn(params: Array[Double],
+                                    l0: Int,
+                                    stepSize: Double): Int = {
+
+    val out = params
+    var l = 0
+    while (!isUTurn(params)) {
+      l += 1
+      steps(1, stepSize)
+      if (l == l0)
+        copy(pqBuf, out)
```

Comment:
`out` is an alias of `params` so you're trashing your params array here which will mess up your acceptance prob computation later. You're going to need some other buffer you can save this version of pqBuf to. I think it's fine to have that buffer as an instance variable of `LeapFrog` that only gets used by this method (as long as you explain that in a comment).